cdap-ui/app/cdap/components/CaskWizards/MicroserviceUpload/MicroserviceQueueEditor.js

```diff
@@ -0,0 +1,184 @@
+  sslTruststoreFilePath: MicroserviceQueueActions.setSSLTruststoreFilePath,
+  sslTruststorePassword: MicroserviceQueueActions.setSSLTruststorePassword,
+  sslTruststoreType: MicroserviceQueueActions.setSSLTruststoreType
+};
+
+const mapDispatchToFieldNameProps = (dispatch, ownProps) => {
+  return {
+    onRemove: () => {
+      dispatch({
+        type: MicroserviceQueueActions.deleteQueue,
+        payload: {index: ownProps.index}
+      });
+    },
+    onAdd: () => {
+      dispatch({
+        type: MicroserviceQueueActions.addQueue,
+        payload: {index: ownProps.index}
+      });
+    },
+    onChange: (fieldProp, e) => {
+      dispatch({
+        type: fieldToActionMap[fieldProp],
+        payload: {
+          index: ownProps.index,
+          [fieldProp]: e.target.value
+        }
+      });
+    }
+  };
+};
+
+let MicroserviceQueueCopy = connect(
+  mapStateToFieldNameProps,
+  mapDispatchToFieldNameProps
+)(MicroserviceQueue);
+
+export default class MicroserviceQueueEditor extends Component {
+  static propTypes = {
+    values: PropTypes.shape({
+      queues: PropTypes.arrayOf(PropTypes.shape({
+        name: PropTypes.string,
+        type: PropTypes.string,
+        properties: PropTypes.object
+      }))
+    }),
+    onChange: PropTypes.func,
+    className: PropTypes.string
+  };
+
+  componentWillMount() {
+    let { values, onChange } = this.props;
+
+    this.state = {
+      queues: values
+    };
+    this.MicroserviceQueueStore = createMicroserviceQueueStore({values});
+
+    this.sub = this.MicroserviceQueueStore.subscribe(() => {
+      let queues = this.MicroserviceQueueStore.getState().general.queues;
+      onChange(queues);
+      this.setState({ queues });
+    });
+
+    this.MicroserviceQueueStore.dispatch({
+      type: MicroserviceQueueActions.onUpdate,
+      payload: {
+        queues: values
+      }
+    });
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.setState({
+      queues: [...nextProps.values]
+    });
+  }
+  shouldComponentUpdate(nextProps) {
+    return this.state.queues.length !== nextProps.values.length;
+  }
+  componentWillUnmount() {
+    this.sub();
+  }
+
+  addFirstQueue = () => {
+    this.MicroserviceQueueStore.dispatch({
+      type: MicroserviceQueueActions.addQueue,
+      payload: { index: 0 }
+    });
+  }
+
+  render() {
+    if (!this.state.queues || !this.state.queues.length) {
+      return (
+        <div
+          className={classnames(this.props.className, "microservice-add-first-queue-container")}
+          onClick={this.addFirstQueue}
+        >
+          <button className="btn add-first-queue-btn btn-link">
+            <IconSVG name="icon-plus" />
+          </button>
+        </div>
+      );
+    }
+    return (
+      <div className="microservice-queue-editor-container">
+        {
+          this.state.queues.map((field, index) => {
+            return (
+              <Provider store={this.MicroserviceQueueStore} key={shortid.generate()}>
```

Comment:
You can move the Provider outside. `Provider` is only for passing the store via context to child components.